src/binder/query/query_graph_label_analyzer.cpp

```diff
@@ -94,49 +95,231 @@ void QueryGraphLabelAnalyzer::pruneNode(const QueryGraph& graph, NodeExpression&
+    table_id_set_t nodeIDs;
+    for (const auto& entry : relEntries) {
+        const auto& relEntry = entry->constCast<RelTableCatalogEntry>();
+        if (direction == RelDataDirection::FWD) {
+            nodeIDs.insert(relEntry.getDstTableID());
+        } else if (direction == RelDataDirection::BWD) {
+            nodeIDs.insert(relEntry.getSrcTableID());
+        } else {
+            KU_UNREACHABLE;
         }
-        for (auto& entry : rel.getEntries()) {
-            auto& relEntry = entry->constCast<RelTableCatalogEntry>();
-            auto srcTableID = relEntry.getSrcTableID();
-            auto dstTableID = relEntry.getDstTableID();
-            if ((srcBoundTableIDSet.contains(srcTableID) &&
-                    dstBoundTableIDSet.contains(dstTableID)) ||
-                (dstBoundTableIDSet.contains(srcTableID) &&
-                    srcBoundTableIDSet.contains(dstTableID))) {
-                prunedEntries.push_back(entry);
+    }
+    return nodeIDs;
+}
+
+std::pair<std::vector<table_id_set_t>, std::vector<table_id_set_t>>
+QueryGraphLabelAnalyzer::pruneRecursiveRel(const std::vector<TableCatalogEntry*>& relEntries,
+    const table_id_set_t srcTableIDSet, const table_id_set_t dstTableIDSet, size_t lowerBound,
+    size_t upperBound, RelDirectionType relDirectionType) const {
+    // src-->[dst,[rels]]
+    std::unordered_map<table_id_t, std::unordered_map<table_id_t, table_id_vector_t>>
+        stepFromLeftGraph;
+    std::unordered_map<table_id_t, std::unordered_map<table_id_t, table_id_vector_t>>
+        stepFromRightGraph;
+    for (auto entry : relEntries) {
+        auto& relEntry = entry->constCast<RelTableCatalogEntry>();
+        auto srcTableID = relEntry.getSrcTableID();
+        auto dstTableID = relEntry.getDstTableID();
+        stepFromLeftGraph[srcTableID][dstTableID].push_back(relEntry.getTableID());
+        stepFromRightGraph[dstTableID][srcTableID].push_back(relEntry.getTableID());
+        if (relDirectionType == RelDirectionType::BOTH) {
+            stepFromLeftGraph[dstTableID][srcTableID].push_back(relEntry.getTableID());
+            stepFromRightGraph[srcTableID][dstTableID].push_back(relEntry.getTableID());
+        }
+    }
+    auto stepFromLeft =
+        pruneRecursiveRel(stepFromLeftGraph, srcTableIDSet, dstTableIDSet, lowerBound, upperBound);
+    auto stepFromRight =
+        pruneRecursiveRel(stepFromRightGraph, dstTableIDSet, srcTableIDSet, lowerBound, upperBound);
+    return {stepFromLeft, stepFromRight};
+}
+
+std::vector<table_id_set_t> QueryGraphLabelAnalyzer::pruneRecursiveRel(
+    const std::unordered_map<table_id_t, std::unordered_map<table_id_t, table_id_vector_t>>& graph,
+    const table_id_set_t& startTableIDSet, const table_id_set_t& endTableIDSet, size_t lowerBound,
+    size_t upperBound) const {
+
+    // There may be multiple rel type between src and dst
+    using PATH = std::vector<table_id_vector_t>;
+    std::vector<std::vector<PATH>> paths(upperBound + 1);
```

Comment:
It would be good to comment 

`std::vector<PATH>` are paths of the same length.

And the outer-most std::vector are paths of different length.

In fact, it might be easier to read if just use `std::vector<PATH>` here.